src/workerd/api/filesystem.h

```diff
@@ -0,0 +1,333 @@
+    JSG_STRUCT(recursive);
+  };
+
+  jsg::Promise<jsg::Ref<FileSystemFileHandle>> getFileHandle(jsg::Lock& js,
+      jsg::USVString name,
+      jsg::Optional<FileSystemGetFileOptions> options,
+      const jsg::TypeHandler<jsg::Ref<jsg::DOMException>>& exception);
+
+  jsg::Promise<jsg::Ref<FileSystemDirectoryHandle>> getDirectoryHandle(jsg::Lock& js,
+      jsg::USVString name,
+      jsg::Optional<FileSystemGetDirectoryOptions> options,
+      const jsg::TypeHandler<jsg::Ref<jsg::DOMException>>& exception);
+
+  jsg::Promise<void> removeEntry(jsg::Lock& js,
+      jsg::USVString name,
+      jsg::Optional<FileSystemRemoveOptions> options,
+      const jsg::TypeHandler<jsg::Ref<jsg::DOMException>>& exception);
+
+  // TODO(node-fs): We are not currently implementing the resolve() method
+  // as defined in the spec. This is a bit tricky as it requires us to work
+  // backwards through the parent directories to find the path to the file.
+  // However, our file node's do not currently have a parent pointer, nor do
+  // they know through which path they are accessible through. Support for this
+  // is not critical so we likely won't implement this before we ship the feature
+  // but it is something we should consider doing in the future.
+  jsg::Promise<kj::Array<jsg::USVString>> resolve(
+      jsg::Lock& js, jsg::Ref<FileSystemHandle> possibleDescendant);
+
+  JSG_ITERATOR(
+      EntryIterator, entries, EntryIteratorType, IteratorState, iteratorNext<EntryIteratorType>);
+  JSG_ITERATOR(KeyIterator, keys, KeyIteratorType, IteratorState, iteratorNext<KeyIteratorType>);
+  JSG_ITERATOR(
+      ValueIterator, values, ValueIteratorType, IteratorState, iteratorNext<ValueIteratorType>);
+
+  void forEach(jsg::Lock& js,
+      jsg::Function<void(
+          jsg::USVString, jsg::Ref<FileSystemHandle>, jsg::Ref<FileSystemDirectoryHandle>)>
+          callback,
+      jsg::Optional<jsg::Value> thisArg);
+
+  JSG_RESOURCE_TYPE(FileSystemDirectoryHandle) {
+    JSG_INHERIT(FileSystemHandle);
+    JSG_METHOD(getFileHandle);
+    JSG_METHOD(getDirectoryHandle);
+    JSG_METHOD(removeEntry);
+    JSG_METHOD(resolve);
+    JSG_METHOD(entries);
+    JSG_METHOD(keys);
+    JSG_METHOD(values);
+    JSG_ITERABLE(entries);
+    JSG_METHOD(forEach);
+  }
+
+ private:
+  kj::Rc<workerd::Directory> inner;
+
+  template <typename Type>
+  static kj::Maybe<Type> iteratorNext(jsg::Lock& js, IteratorState& state) {
+    if (state.index >= state.entries.size()) {
+      return kj::none;
+    }
+
+    auto& entry = state.entries[state.index++];
+
+    if constexpr (kj::isSameType<Type, EntryIteratorType>()) {
+      return EntryType(js.accountedUSVString(entry->getName(js)), entry.addRef());
+    } else if constexpr (kj::isSameType<Type, KeyIteratorType>()) {
+      return js.accountedUSVString(entry->getName(js));
+    } else if constexpr (kj::isSameType<Type, ValueIteratorType>()) {
+      return entry.addRef();
+    } else {
+      KJ_UNREACHABLE;
```

Comment:
this should be a static_assert(false) or similar to denote that it is compile time proven to never be the case.